XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
+// constructed can be accessed as "this". Because init() is called during
+// construction it is not safe to blindly inspect the constructed object at the
+// moment. Though we are currently investigating how to make this safe.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
```

Comment:
>  I don't understand what you mean by "uniquely identify an invocation".

I'll try to be as precise as I can.  lmk if this still isn't clear:  There are two high-level concepts we're talking about here.  We're discussing names, (e.g., handles vs requests).  I'll use _callback_ and _callback invocation_. 

A _callback_  is a "registration" of a javascript function that is to be called asynchronously at some later point during program execution.  It may be invoked 0 to n times.  A _callback_ is uniquely identified by a 53-bit integer called an _async ID_.  It has two lifecycle events - `init()`  & `destroy()`.  

The other concept is a _callback invocation_, which is a specific invocation of a _callback_.  A _callback invocation_ has two lifecycle events `pre()` and `post()`.  Each _callback invocation_ is uniquely identified by the tuple `[async_id, invocation_id]`, where `invocation_id` is simply a count of the number of invocations.